### ISSUE TITLE
Namespace / Directory layout

### ISSUE THREAD

**Issue by clifgriffin** [tommcfarlin/simple-autoloader-for-wordpress#5]
Posting here by your request:

I'm thinking through a plugin namespace/directory layout using your new autoloader.

Typically I'd have my bootstrap file load the main plugin class, and the admin plugin class (in the admin context). 

Any thoughts on where you'd put the main plugin class? 

It's obvious for things like Models and Controllers, and maybe this is indicative that I'm putting too much on a God-class but I'd love your ideas. 

Thanks in advance!

**Comment by clifgriffin** [comment]
Some more thoughts.

I'm thinking a layout like this could work:

App\Main -> main plugin class
App\Admin -> admin only classes
App\Model -> models
App\Controller -> controllers

etc

**Comment by carlalexander** [comment]
I tend to put the main plugin class at the root of the plugin namespace since it's really the "main" class for the entire namespace. Example [here](https://github.com/postmatic/elevated-comments/tree/master/src/CommentIQ).

**Comment by tommcfarlin** [comment]
> Typically I'd have my bootstrap file load the main plugin class, and the admin plugin class (in the admin context).

I tend to use a bootstrap file in a very simple fashion -- usually it'll be on `plugins_loaded` but there are times where other hooks may be sued (or a couple of a classes may be loaded with different hooks.

Generally speaking though, my bootstrap files are now structured:

1. Headers
2. Autoloader
3. Composer Autoloader 
4. Instantiation

> Any thoughts on where you'd put the main plugin class?

I used to have an opinion on this, which I know we've discussed, but what constitutes a "main plugin file" dictates on where I put it.

Some of the things I'm about to share are partially inspired by @carlalexander and @schrapel but I really like the notion of:

* `src`
* `src/API`
* `src/API/Admin`
* `src/Utilities`

etc. Things like `API` and `Utilities` are shared. It's the sub-namespaces (subspaces? :) that'd dictate what goes there.

> It's obvious for things like Models and Controllers, and maybe this is indicative that I'm putting too much on a God-class but I'd love your ideas.

When working on similar stuff, I'm finding that I often _start_ with things that border on god-classes and then I refactor to something smaller. Not always, more often than I'd like to admit.

And I tend to think of models, in WP, as post types or the data around them. Controllers would be what writes data from the model to the database and that retrieves them. There may be even an intermediary layer for sanitizing and validating the data depending on how complicated it really gets. when working with said data.

Hope this helps!
